pkg/scripts/install.go

```diff
@@ -0,0 +1,184 @@
+	InstallOptional(scripts []api.Script, dstDir string) []api.InstallResult
+}
+
+// NewInstaller returns a new instance of the default Installer implementation
+func NewInstaller(image string, scriptsURL string, docker docker.Docker) Installer {
+	return &installer{
+		image:      image,
+		scriptsURL: scriptsURL,
+		docker:     docker,
+		downloader: NewDownloader(),
+		fs:         util.NewFileSystem(),
+	}
+}
+
+type installer struct {
+	image      string
+	scriptsURL string
+	docker     docker.Docker
+	downloader Downloader
+	fs         util.FileSystem
+}
+
+// locationsOrder defines locations in which scripts are searched for in the following order:
+// - script found at the --scripts URL
+// - script found in the application source .sti/bin directory
+// - script found at the default image URL
+var locationsOrder = []string{api.UserScripts, api.SourceScripts, api.DefaultScripts}
+
+// InstallRequired downloads and installs required scripts into dstDir, the result is a
+// map of scripts with detailed information about each of the scripts install process
+// with error if installing some of them failed
+func (i *installer) InstallRequired(scripts []api.Script, dstDir string) (results []api.InstallResult, err error) {
+	results = i.run(scripts, dstDir)
+	failedScripts := []api.Script{}
+	for _, r := range results {
+		if !r.Installed && r.Error != nil {
+			failedScripts = append(failedScripts, r.Script)
+		}
+	}
+	if len(failedScripts) > 0 {
+		err = errors.NewInstallRequiredError(failedScripts)
+	}
+
+	return
+}
+
+// InstallOptional downloads and installs a set of scripts into dstDir, the result is a
+// map of scripts with detailed information about each of the scripts install process
+func (i *installer) InstallOptional(scripts []api.Script, dstDir string) []api.InstallResult {
+	return i.run(scripts, dstDir)
+}
+
+type downloadResult struct {
+	location string
+	err      error
+}
+
+func (i *installer) run(scripts []api.Script, dstDir string) []api.InstallResult {
+	var userResults, sourceResults, defaultResults map[api.Script]*downloadResult
+
+	// get scripts from user provided URL
+	if i.scriptsURL != "" {
+		userResults = i.download(i.scriptsURL, scripts, filepath.Join(dstDir, api.UserScripts))
+	}
+
+	// get scripts from source
+	sourceResults = make(map[api.Script]*downloadResult, len(scripts))
+	for _, script := range scripts {
+		sourceResults[script] = &downloadResult{location: api.SourceScripts}
+		file := filepath.Join(dstDir, api.SourceScripts, string(script))
+		if !i.fs.Exists(file) {
+			sourceResults[script].err = errors.NewDownloadError(file, -1)
+		}
+	}
+
+	// get scripts from default URL
+	defaultURL, err := i.docker.GetScriptsURL(i.image)
+	if err == nil && defaultURL != "" {
+		defaultResults = i.download(defaultURL, scripts, filepath.Join(dstDir, api.DefaultScripts))
+	} else {
+		glog.Warningf("Image %s does not contain STI_SCRIPTS_URL defined!", i.image)
```

Comment:
Maybe it's more accurate to say that you can't get the STI_SCRIPTS_URL from the image (may be defined, but some other error occurred retrieving it)